packages/frontend/core/src/components/affine/page-history-modal/data.ts

```diff
@@ -0,0 +1,213 @@
+import { DebugLogger } from '@affine/debug';
+import {
+  fetchWithTraceReport,
+  type ListHistoryQuery,
+  listHistoryQuery,
+  recoverDocMutation,
+} from '@affine/graphql';
+import { useMutation, useQuery } from '@affine/workspace/affine/gql';
+import { createAffineCloudBlobEngine } from '@affine/workspace/blob';
+import { globalBlockSuiteSchema } from '@affine/workspace/manager';
+import { Workspace } from '@blocksuite/store';
+import { usePageMetaHelper } from '@toeverything/hooks/use-block-suite-page-meta';
+import { useBlockSuiteWorkspacePage } from '@toeverything/hooks/use-block-suite-workspace-page';
+import { revertUpdate } from '@toeverything/y-indexeddb';
+import { useMemo } from 'react';
+import useSWRImmutable from 'swr/immutable';
+import { applyUpdate } from 'yjs';
+
+const logger = new DebugLogger('page-history');
+
+type DocHistory = ListHistoryQuery['workspace']['histories'][number];
+
+export const usePageSnapshotList = (
+  workspaceId: string,
+  pageDocId: string
+): DocHistory[] => {
+  const { data } = useQuery(
+    {
+      query: listHistoryQuery,
+      variables: useMemo(
+        () => ({
+          pageDocId: pageDocId,
+          workspaceId: workspaceId,
+          take: 200,
```

Comment:
u crazy bro

Reply:
I am not sure about it. Maybe change the api to use cursor based pagination?

Reply:
the after param is cursor based.

I don't have enough time to make it standard because of old pagination queries did not follow graphql spec. I may refactor them later.

Reply:
done